extensions/amp-ad-exit/0.1/amp-ad-exit.js

```diff
@@ -0,0 +1,202 @@
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {ClickDelayFilter} from './filters/click-delay';
+import {assertConfig, TransportMode} from './config';
+import {createFilter} from './filters/factory';
+import {isExperimentOn} from '../../../src/experiments';
+import {isJsonScriptTag, openWindowDialog} from '../../../src/dom';
+import {urlReplacementsForDoc} from '../../../src/services';
+import {user} from '../../../src/log';
+
+const TAG = 'amp-ad-exit';
+
+/**
+ * @typedef {{
+ *   finalUrl: string,
+ *   trackingUrls: !Array<string>,
+ *   vars: !./config.Variables,
+ *   filters: !Array<!./filters/filter.Filter>
+ * }}
+ */
+let NavigationTarget;  // eslint-disable-line no-unused-vars
+
+export class AmpAdExit extends AMP.BaseElement {
+  /** @param {!AmpElement} element */
+  constructor(element) {
+    super(element);
+
+    /**
+     * @private @const {!Object<string, !NavigationTarget>}
+     */
+    this.targets_ = {};
+
+    /**
+     * Filters to apply to every target.
+     * @private @const {!Array<!./filters/filter.Filter>}
+     */
+    this.defaultFilters_ = [new ClickDelayFilter('minDelay', 1000)];
+
+    /** @private @struct */
+    this.transport_ = {
+      beacon: true,
+      image: true,
+    };
+
+    this.registerAction('exit', this.exit.bind(this));
+  }
+
+  /**
+   * @param {!../../../src/service/action-impl.ActionInvocation} invocation
+   */
+  exit({args, event}) {
+    user().assert(
+        isExperimentOn(this.win, 'amp-ad-exit'),
+        'amp-ad-exit experiment is off.');
+
+    const target = this.targets_[args.target];
+    user().assert(target, `Exit target not found: '${args.target}'`);
+
+    event.preventDefault();
+    if (!this.filter_(this.defaultFilters_, event) ||
+        !this.filter_(target.filters, event)) {
+      return;
+    }
+    const substituteVariables =
+        this.getUrlVariableRewriter_(args, event, target);
+    if (target.trackingUrls) {
+      target.trackingUrls.map(substituteVariables)
+          .forEach(url => this.pingTrackingUrl_(url));
+    }
+    openWindowDialog(this.win, substituteVariables(target.finalUrl), '_blank');
+  }
+
+
+  /**
+   * @param {!Object<string, string|number|boolean>} args
+   * @param {!../../../src/service/action-impl.ActionEventDef} event
+   * @param {!NavigationTarget} target
+   * @return {function(string): string}
+   */
+  getUrlVariableRewriter_(args, event, target) {
+    const vars = {
+      'CLICK_X': () => event.clientX,
+      'CLICK_Y': () => event.clientY,
+    };
+    const whitelist = {
+      'RANDOM': true,
+      'CLICK_X': true,
+      'CLICK_Y': true,
+    };
+    if (target.vars) {
+      for (const customVar in target.vars) {
+        if (customVar[0] == '_') {
```

Comment:
if the `_` constraint for custom var is only to prevent system vars being overridden, we can do custom var first then let system var to override them.

so that we keep the naming convention across AMP.